src/xml_parser_main_metadata_together.c

```diff
@@ -336,132 +326,213 @@ int cr_xml_parse_main_metadata_together(const char *primary_path,
     assert(other_path);
-    assert(newpkgcb || pkgcb);
     assert(!err || *err == NULL);
 
-    cr_ParserData *primary_pd = NULL;
-    cr_ParserData *filelists_pd = NULL;
-    cr_ParserData *other_pd = NULL;
+    cr_PkgIterator* new_iter = g_new0(cr_PkgIterator, 1);
+    if (!new_iter) {
+        return NULL;
+    }
+
+    new_iter->primary_path = primary_path;
+    new_iter->filelists_path = filelists_path;
+    new_iter->other_path = other_path;
+
+    new_iter->primary_f = NULL;
+    new_iter->filelists_f = NULL;
+    new_iter->other_f = NULL;
+
+    new_iter->primary_pd = NULL;
+    new_iter->filelists_pd = NULL;
+    new_iter->other_pd = NULL;
+
+    new_iter->primary_is_done = 0;
+    new_iter->filelists_is_done = 0;
+    new_iter->other_is_done = 0;
 
-    primary_f = cr_open(primary_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    cr_CbData *cbdata = (cr_CbData*) &new_iter->cbdata;
+    cbdata->in_progress_pkgs_list = NULL;
+    cbdata->in_progress_pkgs_hash = NULL;
+    cbdata->finished_pkgs_list = NULL;
+
+    cbdata->in_progress_count_primary = 0;
+    cbdata->in_progress_count_filelists = 0;
+    cbdata->in_progress_count_other = 0;
+    // Don't think we need these because there's no customization allowed
+    // TODO: maybe whether or not individual files are parsed could be flagged on whether the path is NULL
+    cbdata->newpkgcb = newpkgcb;
+    cbdata->newpkgcb_data = newpkgcb_data;
+
+    new_iter->tmp_err = NULL;
+    new_iter->allow_out_of_order = allow_out_of_order;
+
+    if (allow_out_of_order) {
+        cbdata->in_progress_pkgs_hash = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    }
+
+    GError* tmp_err = new_iter->tmp_err;
+    new_iter->primary_f = cr_open(primary_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
     if (tmp_err) {
-        ret = tmp_err->code;
         g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", primary_path);
         g_clear_error(&tmp_err);
-        goto out;
+        cr_PkgIterator_free(new_iter, err);
+        return NULL;
     }
-    filelists_f = cr_open(filelists_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    new_iter->filelists_f = cr_open(filelists_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
     if (tmp_err) {
-        ret = tmp_err->code;
         g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", filelists_path);
         g_clear_error(&tmp_err);
-        goto out;
+        cr_PkgIterator_free(new_iter, err);
+        return NULL;
     }
-    other_f = cr_open(other_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    new_iter->other_f = cr_open(other_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
     if (tmp_err) {
-        ret = tmp_err->code;
         g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", other_path);
         g_clear_error(&tmp_err);
-        goto out;
+        cr_PkgIterator_free(new_iter, err);
```

Comment:
I think this is not right, the package is given to the user to manage (or to python) it shouldn't be freed here.

Reply:
The user hasn't actually received the pointer yet though, since this is all still in `cr_PkgIterator_new()` - I would think cleaning up and returning `NULL` would be fine?

Reply:
Unless you're actually talking about a different line - calling `pkgcb`.  Because I just found that problem.